### ISSUE TITLE
[Question] Customize tab-completion

### ISSUE THREAD

**Issue by gpotter2** [ipython/ipython#10759]
Hello all,

We're having some troubles with tab-completion on IPython while embeding the IPython shell in our program. Is there any way to fix/customize tabs to fix the two problems we're having below ?

### 1. Pressing [TAB] directly after a slash does not trigger auto completion.

For instance:
![image](https://user-images.githubusercontent.com/10530980/29752659-f480002c-8b62-11e7-97cd-5dfe953ca70a.png)
(note there is no space)
![image](https://user-images.githubusercontent.com/10530980/29752657-ed39908a-8b62-11e7-8e54-015989e1e7d0.png)
(works with a space)

The question for this one is: is there any way we could avoid this ?

### 2. Slashes turn into "path-mode" wrongly

We are using the slash operator to connect objects with each other (using `__div__`).
Is there any way to turn off "path-mode" and force "normal-mode" ? Even better, would it be possible to customize the type of objects that is returned ?

![image](https://user-images.githubusercontent.com/10530980/29752764-3fb5a33e-8b64-11e7-8b52-bc2da8071506.png)

Thanks a lot for any help !

https://github.com/secdev/scapy/pull/703

**Comment by Carreau** [comment]
1) that's weird. I cannot reproduce on master here. I'll try to dig into how you are embeding it. 

2) You should be able to:

Here are the set of matchers that try to find completions. 
There is no easy config options I can see, but by poping the right matcher from this list you should be able to not complete on file names. 

```
In [1]: ip = get_ipython()

In [2]: ip.Completer.matchers
Out[2]:
[<bound method IPCompleter.python_matches of <IPython.core.completer.IPCompleter object at 0x103dbd470>>,
 <bound method IPCompleter.file_matches of <IPython.core.completer.IPCompleter object at 0x103dbd470>>,
 <bound method IPCompleter.magic_matches of <IPython.core.completer.IPCompleter object at 0x103dbd470>>,
 <bound method IPCompleter.python_func_kw_matches of <IPython.core.completer.IPCompleter object at 0x103dbd470>>,
 <bound method IPCompleter.dict_key_matches of <IPython.core.completer.IPCompleter object at 0x103dbd470>>]
```

**Comment by gpotter2** [comment]
My bad:
![image](https://user-images.githubusercontent.com/10530980/29990956-6f1e7c48-8f80-11e7-928f-ede3f1d72042.png)
![image](https://user-images.githubusercontent.com/10530980/29990959-74222474-8f80-11e7-9074-5b6610955638.png)
I was still using IPython 5.

I guess this will never be fixed for Python 2.7, is it ? We're making our soft Python 2.7; 3+ compatible...

**Comment by Carreau** [comment]
> I guess this will never be fixed for Python 2.7, is it ? We're making our soft Python 2.7; 3+ compatible...

It may be, depends how difficult it is to backport, and/or if someone make a PR against 5.x and ask nicely. We are not actively working on 5.x and will push against new features, but if there are annoying bug and someone contribute a fix, we'll probably not say no.
